File: src/main/java/com/weeth/global/auth/kakao/KakaoAuthService.java
package com.weeth.global.auth.kakao;

import com.weeth.global.auth.kakao.dto.KakaoTokenResponse;
import com.weeth.global.auth.kakao.dto.KakaoUserInfoResponse;
import com.weeth.global.config.properties.OAuthProperties;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestClient;

@Service
@Slf4j
public class KakaoAuthService {

    private final OAuthProperties.KakaoProperties kakaoProperties;
    private final RestClient restClient = RestClient.create();

    public KakaoAuthService(OAuthProperties oAuthProperties) {
        this.kakaoProperties = oAuthProperties.getKakao();
    }

    public KakaoTokenResponse getKakaoToken(String authCode) {
        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
        body.add("grant_type", kakaoProperties.getGrantType());
        body.add("client_id", kakaoProperties.getClientId());
        body.add("redirect_uri", kakaoProperties.getRedirectUri());
        body.add("code", authCode);

        return restClient.post()
                .uri(kakaoProperties.getTokenUri())
                .body(body)
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .retrieve()
                .body(KakaoTokenResponse.class);
    }

    public KakaoUserInfoResponse getUserInfo(String accessToken) {
        return restClient.get()
                .uri(kakaoProperties.getUserInfoUri())
                .header("Authorization", "Bearer " + accessToken)
                .retrieve()
                .body(KakaoUserInfoResponse.class);

    }
}

File: src/main/kotlin/com/weeth/global/auth/kakao/KakaoAuthService.kt
package com.weeth.global.auth.kakao

import com.weeth.global.auth.kakao.dto.KakaoTokenResponse
import com.weeth.global.auth.kakao.dto.KakaoUserInfoResponse
import com.weeth.global.config.properties.OAuthProperties
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.client.RestClient
import org.springframework.web.client.body

@Service
class KakaoAuthService(
    oAuthProperties: OAuthProperties,
    restClientBuilder: RestClient.Builder,
) {
    private val kakaoProperties = oAuthProperties.kakao
    private val restClient = restClientBuilder.build()

    fun getKakaoToken(authCode: String): KakaoTokenResponse {
        val body =
            LinkedMultiValueMap<String, String>().apply {
                add("grant_type", kakaoProperties.grantType)
                add("client_id", kakaoProperties.clientId)
                add("redirect_uri", kakaoProperties.redirectUri)
                add("code", authCode)
            }

        return requireNotNull(
            restClient
                .post()
                .uri(kakaoProperties.tokenUri)
                .body(body)
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .retrieve()
                .body<KakaoTokenResponse>(),
        )
    }

    fun getUserInfo(accessToken: String): KakaoUserInfoResponse =
        requireNotNull(
            restClient
                .get()
                .uri(kakaoProperties.userInfoUri)
                .header("Authorization", "Bearer $accessToken")
                .retrieve()
                .body<KakaoUserInfoResponse>(),
        )
}
